File: java/src/main/java/benchmarks/Fasta.java
package benchmarks;

import java.io.ByteArrayOutputStream;
import java.util.*;

public class Fasta extends Benchmark {
    private static final int LINE_LENGTH = 60;

    private static class Gene {
        char ch;
        double prob;

        Gene(char ch, double prob) {
            this.ch = ch;
            this.prob = prob;
        }
    }

    private static final Gene[] IUB = {
        new Gene('a', 0.27), new Gene('c', 0.39), new Gene('g', 0.51),
        new Gene('t', 0.78), new Gene('B', 0.8), new Gene('D', 0.8200000000000001),
        new Gene('H', 0.8400000000000001), new Gene('K', 0.8600000000000001),
        new Gene('M', 0.8800000000000001), new Gene('N', 0.9000000000000001),
        new Gene('R', 0.9200000000000002), new Gene('S', 0.9400000000000002),
        new Gene('V', 0.9600000000000002), new Gene('W', 0.9800000000000002),
        new Gene('Y', 1.0000000000000002)
    };

    private static final Gene[] HOMO = {
        new Gene('a', 0.302954942668), new Gene('c', 0.5009432431601),
        new Gene('g', 0.6984905497992), new Gene('t', 1.0)
    };

    private static final String ALU = 
        "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

    public int n;
    public ByteArrayOutputStream result;

    public Fasta() {
        n = (int) configVal("n");
        result = new ByteArrayOutputStream();
    }

    @Override
    public String name() {
        return "Fasta";
    }

    private char selectRandom(Gene[] genelist) {
        double r = Helper.nextFloat();
        if (r < genelist[0].prob) return genelist[0].ch;

        int lo = 0;
        int hi = genelist.length - 1;

        while (hi > lo + 1) {
            int i = (hi + lo) / 2;
            if (r < genelist[i].prob) {
                hi = i;
            } else {
                lo = i;
            }
        }
        return genelist[hi].ch;
    }

    private void makeRandomFasta(String id, String desc, Gene[] genelist, int n) {
        try {
            result.write((">" + id + " " + desc + "\n").getBytes());

            int todo = n;
            char[] buffer = new char[LINE_LENGTH];

            while (todo > 0) {
                int m = Math.min(todo, LINE_LENGTH);

                for (int i = 0; i < m; i++) {
                    buffer[i] = selectRandom(genelist);
                }

                result.write(new String(buffer, 0, m).getBytes());
                result.write('\n');
                todo -= LINE_LENGTH;
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private void makeRepeatFasta(String id, String desc, String s, int n) {
        try {
            result.write((">" + id + " " + desc + "\n").getBytes());

            int todo = n;
            int k = 0;
            int kn = s.length();

            while (todo > 0) {
                int m = Math.min(todo, LINE_LENGTH);

                while (m >= kn - k) {
                    result.write(s.substring(k).getBytes());
                    m -= kn - k;
                    k = 0;
                }

                if (m > 0) {
                    result.write(s.substring(k, k + m).getBytes());
                    k += m;
                }

                result.write('\n');
                todo -= LINE_LENGTH;
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void run(int iterationId) {
        makeRepeatFasta("ONE", "Homo sapiens alu", ALU, n * 2);
        makeRandomFasta("TWO", "IUB ambiguity codes", IUB, n * 3);
        makeRandomFasta("THREE", "Homo sapiens frequency", HOMO, n * 5);
    }

    @Override
    public long checksum() {
        return Helper.checksum(result.toByteArray());
    }

    public String getResultString() {
        return result.toString();
    }
}

File: scala/src/main/scala/benchmarks/Fasta.scala
package benchmarks

import java.io.ByteArrayOutputStream
import java.util.Locale

class Fasta extends Benchmark:
  private var _n: Int = 0

  def n_=(value: Int): Unit = _n = value
  def n: Int = _n

  private val result = new ByteArrayOutputStream()
  private val LINE_LENGTH = 60

  override def name(): String = "Fasta"

  override def prepare(): Unit =
    _n = configVal("n").toInt

  private case class Gene(ch: Char, prob: Double)

  private val IUB = Array(
    Gene('a', 0.27),
    Gene('c', 0.39),
    Gene('g', 0.51),
    Gene('t', 0.78),
    Gene('B', 0.8),
    Gene('D', 0.8200000000000001),
    Gene('H', 0.8400000000000001),
    Gene('K', 0.8600000000000001),
    Gene('M', 0.8800000000000001),
    Gene('N', 0.9000000000000001),
    Gene('R', 0.9200000000000002),
    Gene('S', 0.9400000000000002),
    Gene('V', 0.9600000000000002),
    Gene('W', 0.9800000000000002),
    Gene('Y', 1.0000000000000002)
  )

  private val HOMO = Array(
    Gene('a', 0.302954942668),
    Gene('c', 0.5009432431601),
    Gene('g', 0.6984905497992),
    Gene('t', 1.0)
  )

  private val ALU =
    "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"

  private def selectRandom(genelist: Array[Gene]): Char =
    val r = Helper.nextFloat()
    if r < genelist(0).prob then return genelist(0).ch

    var lo = 0
    var hi = genelist.length - 1

    while hi > lo + 1 do
      val i = (hi + lo) / 2
      if r < genelist(i).prob then hi = i
      else lo = i

    genelist(hi).ch

  private def makeRandomFasta(id: String, desc: String, genelist: Array[Gene], n: Int): Unit =
    result.write(s">$id $desc\n".getBytes)

    var todo = n
    val buffer = new Array[Char](LINE_LENGTH)

    while todo > 0 do
      val m = math.min(todo, LINE_LENGTH)
      var i = 0
      while i < m do
        buffer(i) = selectRandom(genelist)
        i += 1
      result.write(new String(buffer.take(m)).getBytes)
      result.write('\n'.toInt)
      todo -= LINE_LENGTH

  private def makeRepeatFasta(id: String, desc: String, s: String, n: Int): Unit =
    result.write(s">$id $desc\n".getBytes)

    var todo = n
    var k = 0
    val kn = s.length

    while todo > 0 do
      val m = math.min(todo, LINE_LENGTH)
      var remaining = m

      while remaining >= kn - k do
        result.write(s.substring(k).getBytes)
        remaining -= kn - k
        k = 0

      if remaining > 0 then
        result.write(s.substring(k, k + remaining).getBytes)
        k += remaining

      result.write('\n'.toInt)
      todo -= LINE_LENGTH

  override def run(iterationId: Int): Unit =
    makeRepeatFasta("ONE", "Homo sapiens alu", ALU, _n * 2)
    makeRandomFasta("TWO", "IUB ambiguity codes", IUB, _n * 3)
    makeRandomFasta("THREE", "Homo sapiens frequency", HOMO, _n * 5)

  override def checksum(): Long =
    Helper.checksum(result.toByteArray())

  def getResultString: String = result.toString()
